test/python/test_units.py

```diff
@@ -0,0 +1,561 @@
+        setattr(generic_phase, triple[0], value_1)
+
+    value_3 = [None, None, [1] * generic_phase.n_species]
+    with third_prop_sometimes_fails:
+        setattr(generic_phase, triple[0], value_3)
+
+
+@pytest.fixture
+def pure_fluid_in_vapordome(pure_fluid):
+    pure_fluid.TQ = None, ctu.Q_(0.5, "dimensionless")
+    return pure_fluid
+
+
+def test_set_Q(pure_fluid_in_vapordome):
+    P = pure_fluid_in_vapordome.P
+    T = pure_fluid_in_vapordome.T
+    pure_fluid_in_vapordome.Q = ctu.Q_(0.6, "dimensionless")
+    assert_allclose(pure_fluid_in_vapordome.Q, 0.6 * ctu.units.dimensionless)
+    assert_allclose(pure_fluid_in_vapordome.T, T)
+    assert_allclose(pure_fluid_in_vapordome.P, P)
+
+    pure_fluid_in_vapordome.Q = None
+    assert_allclose(pure_fluid_in_vapordome.Q, 0.6 * ctu.units.dimensionless)
+    assert_allclose(pure_fluid_in_vapordome.T, T)
+    assert_allclose(pure_fluid_in_vapordome.P, P)
+
+    with pytest.raises(ctu.CanteraError, match="an instance of a pint"):
+        pure_fluid_in_vapordome.Q = 0.5
+
+
+def yield_purefluid_only_setters():
+    props = [
+        pytest.param(("TPQ", "T", "P", "Q"), id="TPQ"),
+        pytest.param(("TQ", "T", "Q"), id="TQ"),
+        pytest.param(("PQ", "P", "Q"), id="PQ"),
+        pytest.param(("PV", "P", "v"), id="PV"),
+        pytest.param(("SH", "s", "h"), id="SH"),
+        pytest.param(("ST", "s", "T"), id="ST"),
+        pytest.param(("TH", "T", "h"), id="TH"),
+        pytest.param(("TV", "T", "v"), id="TV"),
+        pytest.param(("VH", "v", "h"), id="VH"),
+        pytest.param(("UP", "u", "P"), id="UP"),
+    ]
+    yield from props
+
+
+def yield_purefluid_only_getters():
+    props = [
+        pytest.param(("DPQ", "density", "P", "Q"), id="DPQ"),
+        pytest.param(("HPQ", "h", "P", "Q"), id="HPQ"),
+        pytest.param(("SPQ", "s", "P", "Q"), id="SPQ"),
+        pytest.param(("SVQ", "s", "v", "Q"), id="SVQ"),
+        pytest.param(("TDQ", "T", "density", "Q"), id="TDQ"),
+        pytest.param(("UVQ", "u", "v", "Q"), id="UVQ"),
+    ]
+    yield from props
+
+
+def yield_all_purefluid_only_props():
+    yield from yield_purefluid_only_setters()
+    yield from yield_purefluid_only_getters()
+
+
+@pytest.mark.parametrize("prop", yield_purefluid_only_setters())
+def test_set_without_units_is_error_purefluid(prop, pure_fluid):
+    value = [None] * (len(prop[0]) - 1) + [0.5]
+    with pytest.raises(ctu.CanteraError, match="an instance of a pint"):
+        # Don't use append here, because append returns None which would be
+        # passed to setattr
+        setattr(pure_fluid, prop[0], value)
+
+
+@pytest.mark.parametrize("props", yield_all_purefluid_only_props())
+def test_multi_prop_getters_purefluid(pure_fluid, props):
+    pair_or_triple, first, second, *third = props
+    first_value, second_value, *third_value = getattr(pure_fluid, pair_or_triple)
+    assert_allclose(getattr(pure_fluid, first), first_value)
```

Comment:
Maybe what would help with making some of these abstracted tests easier to understand is a comment with an example of one case that's being tested, written out in a more literal fashion, e.g.
```
# T, P = pure_fluid.TP
# assert T == pure_fluid.T
# assert P == pure_fluid.P
```
(or at least, that's what I *think* this particular test is doing)